### DISCUSSION TITLE
Ensuring Scalability: Redis Connection Management in Production

### DISCUSSION THREAD

**Discussion by anwar-al-jahwari** [apalis-dev/apalis#434] [Q&A] [answered]
Hello @geofmureithi,

Thank you for creating such an excellent library! I truly appreciate your hard work, and I'm excited to integrate it into our production setup.

![whatsappbot](https://github.com/user-attachments/assets/ca0e7cea-4e22-4fe1-afb6-7ab610f3d278)

To give you a bit more context, I’m building a WhatsApp service expected to handle significant traffic. The architecture involves a webhook that receives messages from Meta servers, which are then routed through a load balancer and a WAF before reaching multiple `/webhook/messages` endpoints. Each of these endpoints enqueues messages as jobs into Redis. We have multiple job processors that then handle these jobs and respond to customers through WhatsApp.

Could you clarify whether the Redis connection in your library uses connection pooling? If not, would this impact performance or scalability in such a high-load environment? I'm particularly interested in ensuring that our Redis interactions remain efficient given the high volume and rapid queuing required.

Thanks again for your fantastic work, and I’d appreciate any insights you could share based on this setup.

**Comment by geofmureithi** [comment, marked as answer]
Hey, good to hear you are checking out apalis and I appreciate your questions they are pretty important.

1. Does apalis support connection pooling in redis? The answer is yes with an asterisk. Basically RedisStorage only cares if you have implemented [aio::ConnectionLike](https://docs.rs/redis/latest/redis/aio/trait.ConnectionLike.html#implementors). We already have an example that  uses deadpool for pooling:
https://github.com/geofmureithi/apalis/blob/f6442fbec34f6a0390363a6cd55118a161e4b7ab/examples/redis-deadpool/src/main.rs#L21-L30
If you want more fine grained approach, feel free to find crates based on the redis crate [here](https://lib.rs/crates/redis/rev) or implement the trait on your own and let me know how it goes.

2. As long as your processing functions are *pure**, you can also scale well horizontally with running several instances of the same binary `worker`. Eg with heroku/dokku::
```
heroku ps:scale web=1 worker=4
```
Let me know if that answers your question.

**Comment by anwar-al-jahwari** [comment]
Thank you so much for your input—I really appreciate it, and I think things are going really well!

Could you explain a bit more about what you mean by the "pure" aspect?

Also, what’s the difference between setting the worker count to 4 versus running 4 separate instances of the worker? Does each worker create a separate connection?

**Reply by geofmureithi** [reply]
> Could you explain a bit more about what you mean by the "pure" aspect?

By pure, I mean a that it doesn't matter which worker takes the job, the result of the processing should be the same given the same inputs. Like in functional programming.

> Also, what’s the difference between setting the worker count to 4 versus running 4 separate instances of the worker?

Currently there is not much difference. The worker count feature will be replaced by concurrency in #428.
This means all you need to set is concurrency of 4 and there will be a max of 4 workers and 4 jobs being processed at any time.

**Reply by geofmureithi** [reply]
That said if you want to provide a pool for each worker, you will have to setup the storage and worker with `Monitor::register` several times. This may improve perf, but I dont believe by much.
If you configure your storage well, ie buffer_size, poll_interval then you should have an optimal experience.

**Reply by anwar-al-jahwari** [reply]
Thank you i will try different approaches and check the metrics and come back with a result if any.
